sql/scheduledlogging: TestCaptureIndexUsageStats failed

sql/scheduledlogging.TestCaptureIndexUsageStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11592440?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11592440?buildTab=artifacts#/) on release-23.1 @ [60e1b100ee2fcb64a9489e75663a3b1116e851ce](https://github.com/cockroachdb/cockroach/commits/60e1b100ee2fcb64a9489e75663a3b1116e851ce):


```
=== RUN   TestCaptureIndexUsageStats
initialized metamorphic constant "parse-json-impl" with value 1
initialized metamorphic constant "mvcc-incremental-iter-tbi" with value false
initialized metamorphic constant "mvcc-value-disable-simple-encoding" with value true
initialized metamorphic constant "storage.value_blocks.enabled" with value false
initialized metamorphic constant "mvcc-max-iters-before-seek" with value 0
initialized metamorphic constant "kv-batch-size" with value 1
initialized metamorphic constant "row-container-rows-per-chunk-shift" with value 1
initialized metamorphic constant "zig-zag-joiner-batch-size" with value 1
initialized metamorphic constant "spilling-queue-initial-len" with value 10
initialized metamorphic constant "merge-joiner-groups-buffer" with value 4
initialized metamorphic constant "ColIndexJoin-using-streamer-batch-size" with value 4367239
initialized metamorphic constant "max-batch-size" with value 9489
initialized metamorphic constant "max-batch-byte-size" with value 17356479
initialized metamorphic constant "resolve-membership-single-scan-enabled" with value false
initialized metamorphic constant "copy-batch-size" with value 44735
initialized metamorphic constant "async-IE-result-channel-buffer-size" with value 15
initialized metamorphic constant "use-index-lookup-for-descriptors-in-database" with value false
initialized metamorphic constant "addsst-rewrite-concurrency" with value 13
initialized metamorphic constant "split-scans-right-for-stats-first" with value true
initialized metamorphic constant "raft-log-truncation-clearrange-threshold" with value 533306
initialized metamorphic constant "kv.expiration_leases_only.enabled" with value true
I230901 18:24:59.339006 1 (gostd) rand.go:249  [-] 1  random seed: -7047880390057109247
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/69cc8b5ff8b35dcd9569961cf48105ef/logTestCaptureIndexUsageStats1163274284
    captured_index_usage_stats_test.go:157: condition failed to evaluate within 45s: expected 8 total entries, got 281
    panic.go:522: -- test log scope end --
--- FAIL: TestCaptureIndexUsageStats (54.58s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cluster-observability
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCaptureIndexUsageStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31157

--- a/pkg/sql/scheduledlogging/BUILD.bazel
+++ b/pkg/sql/scheduledlogging/BUILD.bazel
@@ -36,8 +36,6 @@ go_test(
         "//pkg/security/securityassets",
         "//pkg/security/securitytest",
         "//pkg/server",
-        "//pkg/settings/cluster",
-        "//pkg/testutils",
         "//pkg/testutils/serverutils",
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/testcluster",

--- a/pkg/sql/scheduledlogging/captured_index_usage_stats.go
+++ b/pkg/sql/scheduledlogging/captured_index_usage_stats.go
@@ -71,6 +71,9 @@ type CaptureIndexUsageStatsTestingKnobs struct {
 	// scheduled interval in the case that the logging duration exceeds the
 	// default scheduled interval duration.
 	getOverlapDuration func() time.Duration
+	// onScheduleComplete allows tests to hook into when the current schedule
+	// is completed to check for the expected logs.
+	onScheduleComplete func()
 }
 
 // ModuleTestingKnobs implements base.ModuleTestingKnobs interface.
@@ -147,6 +150,9 @@ func (s *CaptureIndexUsageStatsLoggingScheduler) start(ctx context.Context, stop
 				err := captureIndexUsageStats(ctx, ie, stopper, telemetryCaptureIndexUsageStatsLoggingDelay.Get(&s.st.SV))
 				if err != nil {
 					log.Warningf(ctx, "error capturing index usage stats: %+v", err)
+				}
+				if s.knobs != nil && s.knobs.onScheduleComplete != nil {
+					s.knobs.onScheduleComplete()
 				}
 				dur := s.durationUntilNextInterval()
 				if dur < time.Second {

--- a/pkg/sql/scheduledlogging/captured_index_usage_stats_test.go
+++ b/pkg/sql/scheduledlogging/captured_index_usage_stats_test.go
@@ -21,8 +21,6 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
-	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
-	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -83,34 +81,62 @@ func TestCaptureIndexUsageStats(t *testing.T) {
 	// timeBuffer is a short time buffer to account for non-determinism in the logging timings.
 	const timeBuffer = 4 * time.Second
 
-	settings := cluster.MakeTestingClusterSettings()
-	// Configure capture index usage statistics to be disabled. This is to test
-	// whether the disabled interval works correctly. We start in a disabled
-	// state, once the disabled interval expires, we check whether we have
-	// transitioned to an enabled state, if we have, we check that the expected
-	// logs have been emitted.
-	telemetryCaptureIndexUsageStatsEnabled.Override(context.Background(), &settings.SV, false)
-	// Configure the schedule interval at which we capture index usage
-	// statistics.
-	telemetryCaptureIndexUsageStatsInterval.Override(context.Background(), &settings.SV, stubScheduleInterval)
-	// Configure the schedule interval at which we check whether capture index
-	// usage statistics has been enabled.
-	telemetryCaptureIndexUsageStatsStatusCheckEnabledInterval.Override(context.Background(), &settings.SV, stubScheduleCheckEnabledInterval)
-	// Configure the delay between each emission of index usage stats logs.
-	telemetryCaptureIndexUsageStatsLoggingDelay.Override(context.Background(), &settings.SV, stubLoggingDelay)
+	schedulesFinishedChan := make(chan struct{})
+	logsVerifiedChan := make(chan struct{})
 
 	srv, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
-		Settings: settings,
 		Knobs: base.TestingKnobs{
 			CapturedIndexUsageStatsKnobs: &CaptureIndexUsageStatsTestingKnobs{
 				getLoggingDuration: sd.getLoggingDuration,
 				getOverlapDuration: sd.getOverlapDuration,
+				onScheduleComplete: func() {
+					// Notify that the scheudle has completed.
+					schedulesFinishedChan <- struct{}{}
+					// Wait for logs to be verified before proceeding.
+					<-logsVerifiedChan
+				},
 			},
 		},
 	})
-
 	defer srv.Stopper().Stop(context.Background())
-	s := srv.ApplicationLayer()
+	// Close must be called before the server stops, or we will block indefinitely in the callback above.
+	defer close(logsVerifiedChan)
+
+	numTenants := 1
+	if srv.TenantController().StartedDefaultTestTenant() {
+		numTenants = 2
+	}
+
+	waitForCompletedSchedules := func(count int) {
+		for i := 0; i < count; i++ {
+			<-schedulesFinishedChan
+		}
+	}
+
+	tenantSettings := srv.ApplicationLayer().ClusterSettings()
+
+	// Configure capture index usage statistics to be disabled after the first schedule runs.
+	// This is to test whether the disabled interval works correctly. We will disable the the setting for
+	// the system tenant (if multi-tenant). Because the index usage stats scheduler starts immediately
+	// on server startup, we wait for the first schedule of each tenant to execute before setting all
+	// necessary cluster settings.
+	waitForCompletedSchedules(numTenants)
+	telemetryCaptureIndexUsageStatsEnabled.Override(context.Background(), &tenantSettings.SV, false)
+	telemetryCaptureIndexUsageStatsEnabled.Override(context.Background(), &srv.SystemLayer().ClusterSettings().SV, false)
+	// Configure the schedule interval at which we capture index usage statistics.
+	telemetryCaptureIndexUsageStatsInterval.Override(context.Background(), &tenantSettings.SV, stubScheduleInterval)
+	// Configure the schedule interval at which we check whether capture index
+	// usage statistics has been enabled.
+	telemetryCaptureIndexUsageStatsStatusCheckEnabledInterval.Override(context.Background(), &tenantSettings.SV,
+		stubScheduleCheckEnabledInterval)
+	// Configure the delay between each emission of index usage stats logs.
+	telemetryCaptureIndexUsageStatsLoggingDelay.Override(context.Background(), &tenantSettings.SV, stubLoggingDelay)
+
+	// Allow all tenants to continue.
+	logsVerifiedChan <- struct{}{}
+	if srv.TenantController().StartedDefaultTestTenant() {
+		logsVerifiedChan <- struct{}{}
+	}
 
 	db := sqlutils.MakeSQLRunner(sqlDB)
 
@@ -146,22 +172,32 @@ func TestCaptureIndexUsageStats(t *testing.T) {
 		"index_pkey",
 	}
 
-	// Check that telemetry log file contains all the entries we're expecting, at the scheduled intervals.
-
 	// Enable capture of index usage stats.
-	telemetryCaptureIndexUsageStatsEnabled.Override(context.Background(), &s.ClusterSettings().SV, true)
+	telemetryCaptureIndexUsageStatsEnabled.Override(context.Background(), &tenantSettings.SV, true)
+	//t.Fatal(telemetryCaptureIndexUsageStatsEnabled.Get(&s.ClusterSettings().SV))
+
+	// Check that telemetry log file contains all the entries we're expecting, at the scheduled intervals.
 
 	expectedTotalNumEntriesInSingleInterval := 8
 	expectedNumberOfIndividualIndexEntriesInSingleInterval := 1
 
+	// Wait for channel value from end of 1st schedule.
+	waitForCompletedSchedules(1)
+
 	// Check the expected number of entries from the 1st schedule.
-	testutils.SucceedsSoon(t, func() error {
-		return checkNumTotalEntriesAndNumIndexEntries(t,
-			expectedIndexNames,
-			expectedTotalNumEntriesInSingleInterval,
-			expectedNumberOfIndividualIndexEntriesInSingleInterval,
-		)
-	})
+	require.NoError(t, checkNumTotalEntriesAndNumIndexEntries(t,
+		expectedIndexNames,
+		expectedTotalNumEntriesInSingleInterval,
+		expectedNumberOfIndividualIndexEntriesInSingleInterval,
+	), "error encountered checking the number of total entries and number of index entries")
+
+	// Set the logging duration for the next run to be longer than the schedule
+	// interval duration.
+	stubLoggingDuration := stubScheduleInterval * 2
+	sd.setLoggingDuration(stubLoggingDuration)
+
+	// Allow schedules to proceed.
+	logsVerifiedChan <- struct{}{}
 
 	// Expect number of total entries to hold 2 times the number of entries in a
 	// single interval.
@@ -169,19 +205,18 @@ func TestCaptureIndexUsageStats(t *testing.T) {
 	// Expect number of individual index entries to hold 2 times the number of
 	// entries in a single interval.
 	expectedNumberOfIndividualIndexEntriesAfterTwoIntervals := expectedNumberOfIndividualIndexEntriesInSingleInterval * 2
-	// Set the logging duration for the next run to be longer than the schedule
-	// interval duration.
-	stubLoggingDuration := stubScheduleInterval * 2
-	sd.setLoggingDuration(stubLoggingDuration)
+
+	// Wait for channel value from end of 2nd schedule.
+	waitForCompletedSchedules(1)
 
 	// Check the expected number of entries from the 2nd schedule.
-	testutils.SucceedsSoon(t, func() error {
-		return checkNumTotalEntriesAndNumIndexEntries(t,
-			expectedIndexNames,
-			expectedTotalNumEntriesAfterTwoIntervals,
-			expectedNumberOfIndividualIndexEntriesAfterTwoIntervals,
-		)
-	})
+	require.NoError(t, checkNumTotalEntriesAndNumIndexEntries(t,
+		expectedIndexNames,
+		expectedTotalNumEntriesAfterTwoIntervals,
+		expectedNumberOfIndividualIndexEntriesAfterTwoIntervals,
+	), "error encountered checking the number of total entries and number of index entries")
+
+	logsVerifiedChan <- struct{}{}
 
 	// Expect number of total entries to hold 3 times the number of entries in a
 	// single interval.
@@ -190,17 +225,18 @@ func TestCaptureIndexUsageStats(t *testing.T) {
 	// entries in a single interval.
 	expectedNumberOfIndividualIndexEntriesAfterThreeIntervals := expectedNumberOfIndividualIndexEntriesInSingleInterval * 3
 
+	// Wait for channel value from end of 3rd schedule.
+	waitForCompletedSchedules(1)
+
 	// Check the expected number of entries from the 3rd schedule.
-	testutils.SucceedsSoon(t, func() error {
-		return checkNumTotalEntriesAndNumIndexEntries(t,
-			expectedIndexNames,
-			expectedTotalNumEntriesAfterThreeIntervals,
-			expectedNumberOfIndividualIndexEntriesAfterThreeIntervals,
-		)
-	})
+	require.NoError(t, checkNumTotalEntriesAndNumIndexEntries(t,
+		expectedIndexNames,
+		expectedTotalNumEntriesAfterThreeIntervals,
+		expectedNumberOfIndividualIndexEntriesAfterThreeIntervals,
+	), "error encountered checking the number of total entries and number of index entries")
 
 	// Stop capturing index usage statistics.
-	telemetryCaptureIndexUsageStatsEnabled.Override(context.Background(), &settings.SV, false)
+	telemetryCaptureIndexUsageStatsEnabled.Override(context.Background(), &tenantSettings.SV, false)
 
 	// Iterate through entries, ensure that the timestamp difference between each
 	// schedule is as expected.
@@ -253,6 +289,7 @@ func TestCaptureIndexUsageStats(t *testing.T) {
 		require.Greater(t, actualDuration+timeBuffer, expectedDuration, "%v+%v <= %v", expectedDuration, actualDuration, timeBuffer)
 		previousTimestamp = currentTimestamp
 	}
+
 }
 
 // checkNumTotalEntriesAndNumIndexEntries is a helper function that verifies that

